File: examples/typescript/transaction/src/transaction-state.ts
/**
 * Immutable transaction aggregate state.
 */
export interface TransactionState {
  customerId: string;
  items: LineItem[];
  subtotalCents: number;
  discountCents: number;
  discountType: string;
  status: '' | 'pending' | 'completed' | 'cancelled';
}

export interface LineItem {
  productId: string;
  name: string;
  quantity: number;
  unitPriceCents: number;
}

export function emptyTransactionState(): TransactionState {
  return {
    customerId: '',
    items: [],
    subtotalCents: 0,
    discountCents: 0,
    discountType: '',
    status: '',
  };
}

export function transactionExists(state: TransactionState): boolean {
  return state.status !== '';
}

export function isPending(state: TransactionState): boolean {
  return state.status === 'pending';
}

File: examples/csharp/Transaction/src/TransactionState.cs
using Examples;

namespace Angzarr.Examples.Transaction;

public enum TransactionStatus
{
    New,
    Pending,
    Completed,
    Cancelled
}

/// <summary>
/// Represents the current state of a transaction aggregate.
/// </summary>
public record TransactionState(
    string CustomerId,
    IReadOnlyList<LineItem> Items,
    int SubtotalCents,
    int DiscountCents,
    string DiscountType,
    TransactionStatus Status)
{
    public static TransactionState Empty => new("", Array.Empty<LineItem>(), 0, 0, "", TransactionStatus.New);

    public bool IsPending => Status == TransactionStatus.Pending;
    public bool IsNew => Status == TransactionStatus.New;

    public int CalculateFinalTotal() => Math.Max(0, SubtotalCents - DiscountCents);
    public int CalculateLoyaltyPoints() => CalculateFinalTotal() / 100;
}
